jax/experimental/checkify.py

```diff
@@ -495,16 +564,65 @@ def assert_discharge_rule(error, enabled_errors, pred, code, *, msgs):
 
 ## checkify api
 
-ErrorCategory = enum.Enum('ErrorCategory', ['NAN', 'OOB', 'DIV', 'ASSERT'])
+ErrorCategory = enum.Enum('ErrorCategory', ['NAN', 'OOB', 'DIV', 'USER_CHECK'])
 
-float_errors = {ErrorCategory.NAN, ErrorCategory.DIV}
-index_errors = {ErrorCategory.OOB}
+float_errors = frozenset({ErrorCategory.NAN, ErrorCategory.DIV})
+index_errors = frozenset({ErrorCategory.OOB})
 automatic_errors = float_errors | index_errors
-user_asserts = {ErrorCategory.ASSERT}
+user_checks = frozenset({ErrorCategory.USER_CHECK})
 
 Out = TypeVar('Out')
-def checkify(fun: Callable[..., Out], errors: Set[ErrorCategory] = user_asserts
+
+
+def checkify(fun: Callable[..., Out],
+             errors: FrozenSet[ErrorCategory] = user_checks
              ) -> Callable[..., Tuple[Error, Out]]:
+  """Check for run-time errors in ``fun``.
+
+  Run-time errors are either user-added ``checkify.check`` assertions, or
+  automatically added checks like NaN checks, depending on the ``errors``
+  argument.
+
+  The returned function will return an Error object along with the output of
+  the function. ``err.get()`` will either return ``None`` (if no error occurred)
+  or a string containing an error message. This error message will correspond to
+  the first error which occurred.
+
+  The kinds of errors are:
+    - ErrorCategory.USER_CHECK: a ``checkify.check`` predicate evaluated
+      to False.
+    - ErrorCategory.NAN: a floating-point operation generated a NaN value
+      as output.
+    - ErrorCategory.DIV: division by zero
+    - ErrorCategory.OOB: an indexing operation was out-of-bounds
+
+  Multiple categories can be enabled together (eg. ``errors={ErrorCategory.NAN,
+  ErrorCategory.OOB}``).
+
+  Args:
+    fun: Callable which can contain user checks (see ``check``).
+    errors: Enabled errors. Options are NAN, OOB, DIV. By default USER_CHECK is
+      enabled.
+  Returns:
+    A function with output signature (Error, out)
```

Comment:
```suggestion
    A function which accepts the same arguments as ``fun`` and returns as output
    a pair where the first element is an ``Error`` value, representing any failed
    ``check``s, and the second element is the original output of ``fun``.
```

Reply:
nit: representing any failed ``check``s -> representing the first failed ``check``?